sdk/python/lib/pulumi/x/automation/stack.py

```diff
@@ -0,0 +1,532 @@
+            self.config[key] = ConfigValue(**config[key])
+
+    def __repr__(self):
+        return f"UpdateSummary(result={self.result!r}, version={self.version!r}, start_time='{self.start_time!s} UTC', end_time='{self.end_time!s} UTC')"
+
+
+class BaseResult:
+    stdout: str
+    stderr: str
+    summary: UpdateSummary
+
+    def __init__(self, stdout: str, stderr: str, summary: UpdateSummary):
+        self.stdout = stdout
+        self.stderr = stderr
+        self.summary = summary
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}(result={self.summary.result!r}, start_time='{self.summary.start_time!s} UTC', end_time='{self.summary.end_time!s} UTC')"
+
+
+class UpResult(BaseResult):
+    outputs: OutputMap
+
+    def __init__(self, stdout: str, stderr: str, summary: UpdateSummary, outputs: OutputMap):
+        super().__init__(stdout, stderr, summary)
+        self.outputs = outputs
+
+
+class PreviewResult(BaseResult):
+    pass
+
+
+class RefreshResult(BaseResult):
+    pass
+
+
+class DestroyResult(BaseResult):
+    pass
+
+
+class Stack:
+    name: str
+    """The name identifying the Stack."""
+
+    workspace: Workspace
+    """The Workspace the Stack was created from."""
+
+    @classmethod
+    def create(cls, stack_name: str, workspace: Workspace) -> 'Stack':
+        """
+        Creates a new stack using the given workspace, and stack name.
+        It fails if a stack with that name already exists.
+
+        :param stack_name: The name identifying the Stack
+        :param workspace: The Workspace the Stack was created from.
+        :return: Stack
+        """
+        return Stack(stack_name, workspace, StackInitMode.CREATE)
+
+    @classmethod
+    def select(cls, stack_name: str, workspace: Workspace) -> 'Stack':
+        """
+        Selects stack using the given workspace, and stack name.
+        It returns an error if the given Stack does not exist. All LocalWorkspace operations will call `select` before
+        running.
+
+        :param stack_name: The name identifying the Stack
+        :param workspace: The Workspace the Stack was created from.
+        :return: Stack
+        """
+        return Stack(stack_name, workspace, StackInitMode.SELECT)
+
+    @classmethod
+    def create_or_select(cls, stack_name: str, workspace: Workspace) -> 'Stack':
+        """
+        Tries to create a new stack using the given workspace and stack name if the stack does not already exist,
+        or falls back to selecting the existing stack. If the stack does not exist,
+        it will be created and selected.
+
+        :param stack_name: The name identifying the Stack
+        :param workspace: The Workspace the Stack was created from.
+        :return: Stack
+        """
+        return Stack(stack_name, workspace, StackInitMode.CREATE_OR_SELECT)
+
+    def __init__(self, name: str, workspace: Workspace, mode: StackInitMode) -> None:
+        self.name = name
+        self.workspace = workspace
```

Comment:
Check these args and raise errors if name isn't a string? Workspace isn't a `Workspace`?